napi documentation (#3689)
slint-ui/slint: api/napi/cover.md

<!-- Copyright © SixtyFPS GmbH <info@slint.dev> ; SPDX-License-Identifier: MIT -->
# Slint-node (Beta)

[![npm](https://img.shields.io/npm/v/slint-ui)](https://www.npmjs.com/package/slint-ui)

[Slint](https://slint.dev/) is a UI toolkit that supports different programming languages.
Slint-node is the integration with node.

To get started you can use the [Walk-through tutorial](https://slint.dev/docs/tutorial/node).
We also have a [Getting Started Template](https://github.com/slint-ui/slint-nodejs-template) repository with

> the code of a minimal application using Slint that can be used as a starting point to your program.

Author: Yes I have that in mind. This is the issue now https://github.com/slint-ui/slint/issues/3693

**Warning: Beta**
Slint-node is still in the early stages of development: APIs will change and important features are still being developed.

## Slint Language Manual

The [Slint language manual](../slint) covers the Slint UI description language
in detail.

## Installing Slint

Slint is available via NPM, so you can install by running the following command:

```sh
npm install slint-ui
```

### Dependencies

You need to install the following components:

>   * **[Node.js](https://nodejs.org/download/release/v16.19.1/)** (v16. Newer versions currently not supported: [#961](https://github.com/slint-ui/slint/issues/961))

Author: Yes this hint is no longer necessary. It's removed now.

  * **[npm](https://www.npmjs.com/)**
  * **[Rust compiler](https://www.rust-lang.org/tools/install)** (1.70 or newer)

You will also need a few more dependencies, see <https://github.com/slint-ui/slint/blob/master/docs/building.md#prerequisites>

## Using Slint

To initialize the API, you first need to import the `slint-ui` module in our code:

```js
import * as slint from "slint-ui";
```

As next you can load a slint file with:

```js
let ui = slint.loadFile("ui/main.slint");
```

Combining these two steps leads us to the obligatory "Hello World" example:

```js
import * as slint from "slint-ui";
let ui = slint.loadFile(".ui/main.slint");
let main = new ui.Main();
main.run();
```

See [/examples/todo/node](https://github.com/slint-ui/slint/tree/master/examples/todo/node) for a full example.

## API Overview

### Instantiating a component

The exported component is exposed as a type constructor. The type constructor takes as parameter
an object which allow to initialize the value of public properties or callbacks.

```js
import * as slint from "slint-ui";
// In this example, the main.slint file exports a module which

> // has a counter property and a clicked callback

Reviewer: I suggest to include that file in the example to make it complete. This documentation also talks about the .slint file exporting a module, but that's not the terminology we use. Instead, we say that a .slint file exports components.

I suggest to move this out of the entire code fence and rather have something like this:

~~~
The following example blah blah blah

```
slint code
```

blah blah

```js
sample code but no comments in here unless line specific
```
~~~

Author: done

let ui = slint.loadFile("ui/main.slint");
let component = new ui.MainWindow({
    counter: 42,
    clicked: function() { console.log("hello"); }
});
```

### Accessing a property

> Properties are exposed as properties on the component instance

Reviewer: ```suggestion
Properties declared as `out` or `in-out` in `.slint` files are visible as JavaScript on the component instance.
```

(minor note: It's not explained here what a component instance is, nor is there a link to docs that explain it for this API)

```js
component.counter = 42;
console.log(component.counter);
```

### Callbacks

> The callbacks are also exposed as property that have a setHandler function, and that can can be called.

Reviewer: ```suggestion
The callbacks are also exposed as property that has a setHandler function, and that can be called as a function.
```

Please complete the example with .slint code, so that the example is self-contained.

> The callbacks are also exposed as property that has a setHandler function, and that can be called as a function.

Reviewer: Since the suggestion was marked as resolved but my associated comment disappeared, I'll add it again:

Please complete the example with .slint code, so that the example is self-contained.

```js
// connect to a callback
component.clicked.setHandler(function() { console.log("hello"); })
// emit a callback
component.clicked();
```

### Type Mappings

> 

Reviewer: Please include an intro sentence here, like we have in the Rust documentation.

> | `.slint` Type | JavaScript Type | Notes |

Author: done

> | --- | --- | --- |

Reviewer: This table is also missing an entry for images.

Author: It's now part of the list.

| `int` | `Number` | |
| `float` | `Number` | |
| `string` | `String` | |
| `color` | `Color` |  |
| `brush` | `Brush` |  |
| `length` | `Number` | |
| `physical_length` | `Number` | |
| `duration` | `Number` | The number of milliseconds |

> | `angle` | `Number` | The value in degrees |

Reviewer: As per commit e0d1d4366d03e4a39b0888c9c1448a9fc10eed2e

```suggestion
| `angle` | `Number` | The angle in degrees |
```

> | structure | `Object` | Structures are mapped to JavaScrip objects with structure fields mapped to properties. |

Reviewer: ```suggestion
| structure | `Object` | Structures are mapped to JavaScript objects where each structure field is a property. |
```

> | array | `Array` or Model Object | |

Reviewer: Instead of "Model Object", shouldn't it perhaps be "an implementation of the Model<T>" interface (where Model<T> is rendered as a link to the `Model<T>` interface) ?

> ### Models

Reviewer: Let's use the same heading as in the Slint Language Documentation

```suggestion
### Arrays and Models
```

> For property of array type, they can either be set using an array.
> In that case, getting the property also return an array.
> If the array was set within the .slint file, the array can be obtained

Reviewer: ```suggestion
[Array properties](../slint/src/reference/types#arrays-and-models) can be set from JavaScript by passing
either `Array` objects or implementations of the {@link Model} interface.

When passing a JavaScript `Array` object, the contents of the array are copied. Any changes to the JavaScript afterwards will not be visible on the Slint side. Similarly, reading a Slint array property from JavaScript that was
previously initialised from a JavaScript `Array`, will return a newly allocated JavaScript `Array`.
```

```js
component.model = [1, 2, 3];

> // component.model.push(4); // does not work, because it operate on a copy

Reviewer: ```suggestion
// component.model.push(4); // does not work, because assignment creates a copy.
```

> // but re-assigning works

Reviewer: ```suggestion
// Use re-assignment instead.
```

component.model = component.model.concat(4);
```

> Another option is to set a model object.  A model object has the following function:

Reviewer: ```suggestion
Another option is to set an object that implements the {@link Model} interface. Rreading a Slint array property from JavaScript that was previously initialised from a {@link Model} object, will return a reference to the model. 
```

> * `rowCount()`: returns the number of element in the model.
> * `rowData(index)`: return the row at the given index
> * `setRowData(index, data)`: called when the model need to be changed. `this.notify.rowDataChanged` must be called if successful.
> 
> When such an object is set to a model property, it gets a new `notify` object with the following function
> 
> * `rowDataChanged(index)`: notify the view that the row was changed.
> * `rowAdded(index, count)`: notify the view that rows were added.
> * `rowRemoved(index, count)`: notify the view that a row were removed.
> * `reset()`: notify the view that everything may have changed.
> 

Reviewer: I think these belong into the type documentation of `Model<T>` and `ModelPeer<T>`, referencing each other.

```suggestion
```

> As an example, here is the implementation of the `ArrayModel` (which is available as `slint.ArrayModel`)

Reviewer: ```suggestion
As an example, here is the implementation of the @{link ArrayModel}.
```

> ```js

Reviewer: I feel that similar to how the Model docs in Rust contain the VecModel as example, the Model<T> docs should have the Array Model as example, not here.

Author: done
